🧨 💎 Add literal interaction and outline for tests

cthoyt opened the pull request:
Closes #244. This PR is part of the break-up effort on #107.

This PR implements literal interaction modules, which wrap another interaction module as well as a "combination" function, which takes the representations for the entities and representations for literals and creates a combined representation (implementation up to user - the simplest is to concatenate them).  

There's a bit of goofiness to the typing since it's hard to express types dynamically - this makes it difficult to think about any other kind of interaction function besides simple ones that give single embeddings for head, relation, and tail.

## Before Merge

- [x] Make tests work!
- [x] Do we need regularizers for LiteralE in general, or for DistMultLiteral or ComplExLiteral?
- [x] Should we reorganize the `Combination` class to take the unconcatenated representations?

cthoyt committed:
Add literal interaction and outline for tests

References #244

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -311,6 +312,51 @@ def reset_parameters(self):
+        h = torch.cat(h, dim=-1)
+        h = self.combination(h.view(-1, h.shape[-1])).view(*h.shape[:-1], -1)  # type: ignore
+        t = torch.cat(t, dim=-1)
+        # TODO what if tail representation is not a single torch.FloatTensor?
```
Below this line `t` is a tensor. The actual question arises one line above: why do we always concat before?

cthoyt replied:
@mberr good question. can you look into this?

cthoyt replied:
I know the answer now. it's because h is a tuple of two tensors for the actual embedding then the literal embedding. Maybe it would be better to make the combination take them as two separate tensors and do its work on them rather than concatenating, then splitting apart again

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Add base class for literal models

cthoyt committed:
Add new models for LiteralE

Trigger CI

cthoyt committed:
Add tests for literal embeddings

cthoyt committed:
Reorganize LiteralE base class

Since it needs to auto-append the literal representation, it can't do magic stuff in the middle

cthoyt committed:
Update base.py

cthoyt commented on tests/test_nn/test_modules_literal.py:
```diff
@@ -0,0 +1,47 @@
+nations_literal = NationsLiteral()
+
+
+class MockLiteralInteraction(LiteralInteraction):
```
@mberr couldnyou help look into why these tests aren't passing?

mberr replied:
@cthoyt now some tests do not pass since the `LiteralInteraction` is not a subclass of `FunctionalInteraction` and thus does not have `_prepare_for_functional`

<summary>Traceback</summary>
<details>

```python-traceback
tests/cases.py:457 (LiteralTests.test_forward_consistency_with_functional)
self = <tests.test_nn.test_modules_literal.LiteralTests testMethod=test_forward_consistency_with_functional>

    def test_forward_consistency_with_functional(self):
        """Test forward's consistency with functional."""
        # set in eval mode (otherwise there are non-deterministic factors like Dropout
        self.instance.eval()
        for hs, rs, ts in self._get_test_shapes():
            h, r, t = self._get_hrt(hs, rs, ts)
            scores = self.instance(h=h, r=r, t=t)
>           kwargs = self.instance._prepare_for_functional(h=h, r=r, t=t)

../cases.py:465: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = LiteralInteraction(
  (base): ComplExInteraction()
  (combination): Sequential(
    (0): Linear(in_features=4, out_features=2, bias=True)
    (1): Dropout(p=0.1, inplace=False)
  )
)
name = '_prepare_for_functional'

    def __getattr__(self, name: str) -> Union[Tensor, 'Module']:
        if '_parameters' in self.__dict__:
            _parameters = self.__dict__['_parameters']
            if name in _parameters:
                return _parameters[name]
        if '_buffers' in self.__dict__:
            _buffers = self.__dict__['_buffers']
            if name in _buffers:
                return _buffers[name]
        if '_modules' in self.__dict__:
            modules = self.__dict__['_modules']
            if name in modules:
                return modules[name]
>       raise AttributeError("'{}' object has no attribute '{}'".format(
            type(self).__name__, name))
E       AttributeError: 'LiteralInteraction' object has no attribute '_prepare_for_functional'

../../../../venv/pykeen/lib/python3.8/site-packages/torch/nn/modules/module.py:947: AttributeError
```
</details>

cthoyt replied:
is there a reason we need these? do we need to modify these tests, or can I just override them to not test anything?

mberr replied:
I suppose we do not need them. :sweat_smile:

cthoyt replied:
oky I think I solved it

cthoyt replied:
in the monotonic affine transform class, these were just clobbered, but I added some checks inside the parent class itself to automatically skip these tests if not possible

mberr committed:
Fix test tensor creation

mberr committed:
some fixes to testing

cthoyt commented on src/pykeen/models/multimodal/combinations.py:
```diff
@@ -0,0 +1,62 @@
+        input_dropout: float = 0.0,
+    ):
+        linear = nn.Linear(embedding_dim + num_of_literals, embedding_dim)
+        dropout = nn.Dropout(input_dropout)
```
@mberr note the dropout comes after the linear later for distmult but before for complex. was this correct?

mberr committed:
Merge remote-tracking branch 'origin/master' into add-literal-interaction-modules

# Conflicts:
#	src/pykeen/models/multimodal/complex_literal.py
#	src/pykeen/models/multimodal/distmult_literal.py

mberr committed:
update imports

cthoyt committed:
Update combinations.py

cthoyt commented on src/pykeen/models/multimodal/complex_literal.py:
```diff
@@ -45,74 +48,42 @@ def __init__(
+                    embedding_dim=embedding_dim,
+                    initializer=nn.init.xavier_normal_,
+                    dtype=torch.complex64,
+                    # TODO: verify
```
this needs to be written up a bit differently - no regularization was reported for this model, right? so it's time to remove it

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Reorganize handling of literal tensor

cthoyt committed:
Pass flake8

cthoyt committed:
Fix testing for non-functional

Trigger CI

cthoyt committed:
Remove unnecessary variable passing

cthoyt committed:
Reorganize combinations code

cthoyt committed:
Fix typing

cthoyt committed:
Update combinations.py

cthoyt committed:
Update combinations.py

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Update combinations.py

cthoyt committed:
Fix tests

Trigger CI

cthoyt committed:
Reorganize

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Add tests for both distmult and complex

@mberr right now there's an issue with the complEx one

mberr committed:
Add fixme

cthoyt committed:
Address FIXME

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Update remaining docs and imports

Also remove dead class MultimodalModel

cthoyt committed:
Fix math

Trigger CI

cthoyt committed:
Use kwargs to pass information to base model

cthoyt committed:
Update resolving of interaction in literal interaction

cthoyt committed:
Pass flake8

Trigger CI

mberr commented on src/pykeen/nn/combinations.py:
```diff
@@ -0,0 +1,93 @@
+        input_dropout: float = 0.0,
+    ):
+        real = nn.Sequential(
+            nn.Dropout(input_dropout),
```
This inconsistency in the order of linear layer and dropout was from the original implementations, right?

cthoyt replied:
yes, I mentioned it in an earlier comment. If we want to update it, now would be a good time

mberr replied:
Hm, I have some plans to add a dropout option to the Embedding class itself, which would make the dropout then linear version unnecessary. So maybe we can go consistently with the linear then dropout variant?

mberr replied:
Here is the link to the earlier comment: https://github.com/pykeen/pykeen/pull/245#discussion_r595197755

cthoyt replied:
Handled in c7b9bc19

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -326,6 +328,52 @@ def reset_parameters(self):
+        combination: Combination,
+        base_kwargs: Optional[Mapping[str, Any]] = None,
+    ):
+        # TODO documentation for this
```
todo :slightly_smiling_face:

cthoyt replied:
😅

cthoyt replied:
Done in c7b9bc19

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -326,6 +328,52 @@ def reset_parameters(self):
+        super().__init__()
+        self.base = interaction_resolver.make(base, base_kwargs)
+        self.combination = combination
+        self.entity_shape = tuple(self.base.entity_shape) + ("e",)
```
Do we need to make sure that `"e"` is unused?

cthoyt replied:
yes, there needs to be an extra dimension floating around here that gets added in by the literal model

cthoyt replied:
Added more docs for this in c7b9bc19

mberr commented on tests/cases.py:
```diff
@@ -462,10 +465,14 @@ def test_scores(self):
+                scores_f = self.cls.func(**kwargs).view(-1)
+            else:
+                kwargs = dict(h=h, r=r, t=t)
+                scores_f = self.instance(h=h, r=r, t=t)
```
no need for `view(-1)` here?

cthoyt replied:
it works without it... sooooooo... you tell me

mberr reviewed (approved):
besides, looks good to me! Great to have them integrated into testing.

cthoyt committed:
Merge branch 'master' into add-literal-interaction-modules

cthoyt committed:
Update minimum pystow

Closes #399

cthoyt committed:
Reorganize code and update docs

Trigger CI
